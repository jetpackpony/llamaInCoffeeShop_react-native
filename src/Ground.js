import React from 'react';
import {
  StyleSheet,
  View,
  TouchableWithoutFeedback
} from 'react-native';

const times = (n, callback) => {
  let arr = [];
  for(let i = n; i > 0; i--) {
    arr.push(i);
  }
  return arr.reverse().map(callback);
};


export default ({
  position, width, tileWidth, groundHeight
}) => {
  const tilesNum = Math.ceil(width / tileWidth) + 1;
  const tiles = times(tilesNum, (i) => (
    <View key={i} style={styles.tile}></View>
  ));
  return (
    <View style={[styles.container, { left: position, height: groundHeight }]}>
      {tiles}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    backgroundColor: 'green',
    bottom: 0,
    left: 0,

    flex: 1,
    flexDirection: 'row',
  },
  tile: {
    width: 100,
    backgroundColor: 'red',
    borderWidth: 1,
    borderColor: 'black'
  }
});
