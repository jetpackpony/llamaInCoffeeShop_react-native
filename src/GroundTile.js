import React from 'react';
import {
  StyleSheet,
  View,
  Image
} from 'react-native';
import grass from './sprites/grassBlock.png';

export default ({ width }) => {
  return (
    <View style={{ width }}>
      <Image style={styles.image} source={grass} />
    </View>
  );
};

const styles = StyleSheet.create({
  image: {
    position: 'absolute',
    top: -50
  }
});
