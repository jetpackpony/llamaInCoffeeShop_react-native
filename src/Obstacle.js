import React from 'react';
import {
  StyleSheet,
  View
} from 'react-native';

export default ({ obj, groundHeight }) => {
  return (
    <View style={[
      styles.obstacle,
      { bottom: groundHeight, left: obj.coords.x }
    ]}></View>
  );
};

const styles = StyleSheet.create({
  obstacle: {
    position: 'absolute',
    width: 50,
    height: 50,
    backgroundColor: 'red'
  }
});
