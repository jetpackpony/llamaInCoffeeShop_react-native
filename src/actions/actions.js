export const ActionTypes = {
  JUMP_START: 'JUMP_START',
  TICK: 'TICK'
};

export function jumpStart() {
  return { type: ActionTypes.JUMP_START };
}

export function tick(timestamp) {
  return { type: ActionTypes.TICK, payload: { timestamp } };
}
