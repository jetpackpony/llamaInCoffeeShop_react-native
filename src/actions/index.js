export const ActionTypes = {
  JUMP: 'JUMP',
  TICK: 'TICK',
  SET_DIMENSIONS: 'SET_DIMENSIONS'
};

export function jump() {
  return { type: ActionTypes.JUMP };
}

export function tick(timestamp) {
  return { type: ActionTypes.TICK, payload: { timestamp } };
}

export function setDimensions(width, height) {
  return { type: ActionTypes.SET_DIMENSIONS, payload: { width, height } };
}
