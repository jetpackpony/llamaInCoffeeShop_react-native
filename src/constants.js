export const GRAVITY = -5000;
export const GROUND_Y = 80;
export const WORLD_SPEED = -300;
