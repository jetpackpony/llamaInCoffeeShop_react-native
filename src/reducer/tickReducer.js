const speed = 0.05; // pixels per milisecond
const performTick = (state, timestamp) => {
  // If the first frame, just set the new timestamp
  if (state.lastTick === 0) {
    return { ...state, lastTick: timestamp };
  }

  // Calculate number of pixels to move
  const frameLag = timestamp - state.lastTick;
  const pixToMove = frameLag * speed;

  // If less than one pixel, skip this frame
  if (pixToMove < 1) {
    return state;
  }

  // Else, update coordinates and timestamp
  return {
    ...state,
    coord: state.coord + pixToMove,
    lastTick: timestamp
  };
};

export default function tickReducer(state, action) {
  return {
    ...state,
    player: {
      ...state.player,
      ...performTick(state.player, action.payload.timestamp)
    }
  };
};
