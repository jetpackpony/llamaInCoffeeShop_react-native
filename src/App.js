import React, { Component } from 'react';
import {
  StyleSheet,
  View
} from 'react-native';
import { createStore } from 'redux';
import Reducer from './reducer/Reducer';
import Player from './Player';
import { jumpStart } from './actions/Actions';

const store = createStore(Reducer);

export default class LlamaInCoffeeShop extends Component {
  render() {
    return (
      <View style={styles.container}>
        <Player
          position={store.getState().player.coord}
          onPressIn={() => {
            console.log('pressed');
            store.dispatch(jumpStart());
            console.log("store:", store.getState());
          }}
        />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'flex-end',
    alignItems: 'center',
    padding: 100
  }
});
