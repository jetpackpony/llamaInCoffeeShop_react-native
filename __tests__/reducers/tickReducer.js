import tickReducer from '../../src/reducer/tickReducer';
import { tick } from '../../src/actions';

const getInitialState = () => ({
  player: {
    coord: 10,
    inJump: false,
    lastTick: 0
  }
});

it('updates lastTick when lastTick is zero', () => {
  expect(
    tickReducer(getInitialState(), tick(100))
  ).toMatchSnapshot();
});

it('updates player\'s position', () => {
  let state = getInitialState();
  state.player.lastTick = 1000;

  expect(
    tickReducer(state, tick(2000))
  ).toMatchSnapshot();
});

it('skips a frame if the player position moved less than a pixel', () => {
  let state = getInitialState();
  state.player.lastTick = 1000;

  expect(
    tickReducer(state, tick(1019))
  ).toMatchSnapshot();
});
