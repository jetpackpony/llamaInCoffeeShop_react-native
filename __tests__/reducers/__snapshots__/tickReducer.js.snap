// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`skips a frame if the player position moved less than a pixel 1`] = `
Object {
  "player": Object {
    "coord": 10,
    "inJump": false,
    "lastTick": 1000,
  },
}
`;

exports[`updates lastTick when lastTick is zero 1`] = `
Object {
  "player": Object {
    "coord": 10,
    "inJump": false,
    "lastTick": 100,
  },
}
`;

exports[`updates player's position 1`] = `
Object {
  "player": Object {
    "coord": 60,
    "inJump": false,
    "lastTick": 2000,
  },
}
`;
